test/arbitrage/BancorArbitrage.ts

```diff
@@ -965,12 +978,79 @@ describe('BancorArbitrage', () => {
 
-                // expect to approve exactly the amounts needed for the second trade for each exchange
-                await expect(bancorArbitrage.connect(user).execute(routes, AMOUNT))
-                    .to.emit(arbToken2, 'Approval')
-                    .withArgs(bancorArbitrage.address, exchanges.address, secondHopSourceAmount);
+                // expect to approve MAX_UINT256 for the exchange
+                const approveAmount = MAX_UINT256;
+                const approveExchange = exchangeId === ExchangeId.BancorV3 ? network.address : exchanges.address;
+                const allowance = await arbToken1.allowance(bancorArbitrage.address, approveExchange);
+                if(allowance === 0) {
+                    await expect(bancorArbitrage.connect(user).execute(routes, AMOUNT))
+                        .to.emit(arbToken1, 'Approval')
+                        .withArgs(bancorArbitrage.address, approveExchange, approveAmount);
+                }
+            }
+        });
+
+        it('should be exempt from trading fees on Bancor V3', async () => {
+            const { token: arbToken1 } = await prepareBancorV3PoolAndToken(tokenSymbols[0]);
+            const { token: arbToken2 } = await prepareBancorV3PoolAndToken(tokenSymbols[1]);
+            const { token: eth } = await prepareBancorV3PoolAndToken(tokenSymbols[2]);
+
+            // transfer tokens to mock exchanges
+            await transfer(deployer, bnt, exchanges.address, AMOUNT.mul(20));
+            await transfer(deployer, arbToken1, exchanges.address, AMOUNT.mul(20));
+            await transfer(deployer, arbToken2, exchanges.address, AMOUNT.mul(20));
+
+            for(const token of [arbToken2, eth]) {
+                const routes = [
+                    {
+                        exchangeId: ExchangeId.BancorV2,
+                        targetToken: arbToken1.address,
+                        minTargetAmount: 1,
+                        deadline: DEADLINE,
+                        customAddress: arbToken1.address,
+                        customInt: 0
+                    },
+                    {
+                        exchangeId: ExchangeId.BancorV3,
+                        targetToken: token.address,
+                        minTargetAmount: 1,
+                        deadline: DEADLINE,
+                        customAddress: token.address,
+                        customInt: 0
+                    },
+                    {
+                        exchangeId: ExchangeId.BancorV2,
+                        targetToken: bnt.address,
+                        minTargetAmount: 1,
+                        deadline: DEADLINE,
+                        customAddress: bnt.address,
+                        customInt: 0
+                    }
+                ];
+
+                // check that the fee amount is 0
+                // use different fees
+                for (const tradeFee of [0.03, 0.05, 1, 2, 3]) {
```

Comment:
Watch out as `0.03` doesn't have an accurate representation AFAIK.
Might be better pass an ASCII (string) representation of it.

Reply:
Outputted the resulting trade fee for each item here and the resulting PPM conversion- it seems okay.